music_assistant/server/providers/plex/__init__.py

```diff
@@ -0,0 +1,467 @@
+        """Get full playlist details by id."""
+        plex_playlist = await self._get_data(prov_playlist_id, PlexPlaylist)
+        return await self._parse_playlist(plex_playlist)
+
+    async def get_playlist_tracks(  # type: ignore[return]
+        self, prov_playlist_id: str
+    ) -> AsyncGenerator[Track, None]:
+        """Get all playlist tracks for given playlist id."""
+        plex_playlist = await self._get_data(prov_playlist_id, PlexPlaylist)
+
+        playlist_items = await self._run_async(plex_playlist.items)
+
+        if not playlist_items:
+            yield None
+        for index, plex_track in enumerate(playlist_items):
+            track = await self._parse_track(plex_track)
+            if track:
+                track.position = index + 1
+                yield track
+
+    async def get_artist_albums(self, prov_artist_id) -> list[Album]:
+        """Get a list of albums for the given artist."""
+        plex_artist = await self._get_data(prov_artist_id, PlexArtist)
+        plex_albums = await self._run_async(plex_artist.albums)
+        if plex_albums:
+            albums = []
+            for album_obj in plex_albums:
+                albums.append(await self._parse_album(album_obj))
+            return albums
+        return []
+
+    async def get_stream_details(self, item_id: str) -> StreamDetails | None:
+        """Get streamdetails for a track."""
+        plex_track = await self._get_data(item_id, PlexTrack)
+        if not plex_track or not plex_track.media:
+            raise MediaNotFoundError(f"track {item_id} not found")
+
+        media: PlexMedia = plex_track.media[0]
+        media_part: PlexMediaPart = media.parts[0]
+        audio_stream: PlexAudioStream = media_part.audioStreams()[0]
+
+        media_type = ContentType.try_parse(media.container)
+
+        stream_details = StreamDetails(
+            item_id=plex_track.key,
+            provider=self.domain,
+            content_type=ContentType.try_parse(media.container),
+            duration=plex_track.duration,
+            channels=media.audioChannels,
+            data=plex_track,
+            loudness=audio_stream.loudness,
+        )
+
+        if media_type != ContentType.M4A:
+            stream_details.direct = self._plex_server.url(media_part.key, True)
+            if audio_stream.samplingRate:
+                stream_details.sample_rate = audio_stream.samplingRate
+            if audio_stream.bitDepth:
+                stream_details.bit_depth = audio_stream.bitDepth
+
+        else:
+            url = plex_track.getStreamURL()
+            media_info = await parse_tags(url)
+
+            stream_details.channels = media_info.channels
+            stream_details.content_type = ContentType.try_parse(media_info.format)
+            stream_details.sample_rate = media_info.sample_rate
+            stream_details.bit_depth = media_info.bits_per_sample
+
+        return stream_details
```

Comment:
if you have just an url to stream, you'd better set it as "direct" attribute of the streamdetails so url is fed directly into ffmpeg, saves you form having to chunk the audio yourself with the get_audio_stream

Reply:
If it is not an m4a, I set the direct attribute. If it is an m4a, we use the Plex transcoding and in that case the seek position has to be included in the URL, which I build in the get_audio_stream method. And there I have to return the binary data like a do it right now, haven't I? Is there a better way to solve that?

Reply:
Oh sorry, in that case I misread the code and its all good now.